### PR TITLE
Seed 2.3.1 cleanup

### PR DESCRIPTION
**Pull request by thorsent**
A mid-level cleanup of the seed project. Increases error checking, colored output, removed dead code, minor updates to presentation components, added optional proxy server, fixed inconsistent capitalization.

This should be a drop in replacement. Compatible with Finsemble 2.3

### CODE REVIEW

**Comment by thorsent** [comment]
@kriswest @NinjaManatee please just give this branch a quick test to make sure I haven't screwed anything up.

**Comment by thorsent** [comment]
@carterbs I merged the copy file changes into this branch but they don't seem to be working correctly. Changes I make to my component don't make it into dist anymore.

File: configs/application/config.json [hunk 1]
```diff
@@ -1,15 +1,13 @@
...
-	"isAuthEnabled": false,
 	"storage": {
 		"topicToDataStoreAdapters" : {
 			"finsemble" : "LocalStorageAdapter",
 			"finsemble.workspace" : "LocalStorageAdapter",
 			"finsemble.workspace.cache" : "LocalStorageAdapter"
 		},
 		"dataStoreAdapters" : {
-			"LocalStorageAdapter": "$applicationRoot/adapters/LocalStorageAdapter.js"
+			"LocalStorageAdapter": "/adapters/LocalStorageAdapter.js"
```
Comment (NinjaManatee) [inline configs/application/config.json]: Should `$applicationRoot` be removed from this? Is this combined with applicationRoot when read from the configuration?

File: gulpfile.js [hunk 2]
```diff
@@ -22,28 +24,39 @@
 	const extensions = fs.existsSync("./gulpfile-extensions.js") ? require("./gulpfile-extensions.js") : undefined;
 	const async = require("async");
```
Comment (NinjaManatee) [inline gulpfile.js]: I think this should be moved up with the other require imports.

File: gulpfile.js [hunk 3]
```diff
@@ -22,28 +24,39 @@
...
+			src: ['configs/**/*'],
+			dest: 'configs/',
+		},
+		{
+			src: ['node_modules/@chartiq/finsemble/dist/**/*'],
+			dest: path.join(__dirname, "Finsemble/")
+		}
+	];
 	// #region Constants
 	const startupConfig = require("./configs/other/server-environment-startup");
```
Comment (NinjaManatee) [inline gulpfile.js]: I think this should be moved up next to the gulpfile-extensions import.

File: gulpfile.js [hunk 4]
```diff
@@ -22,28 +24,39 @@
...
 	//Force colors on terminals.
 	let angularComponents;
 	try {
 		angularComponents = require("./build/angular-components.json");
```
Comment (NinjaManatee) [inline gulpfile.js]: This import should be done at the top of the file.

File: gulpfile.js [hunk 5]
```diff
@@ -111,6 +124,7 @@
 				path.join(srcPath, "components", "**", "*.scss"),
 				path.join(__dirname, "src-built-in", "components", "**", "*.scss"),
 			];
+
 			// // Don't build files built by angular
```
Comment (NinjaManatee) [inline gulpfile.js]: Should this commented out code be removed?

File: gulpfile.js [hunk 6]
```diff
@@ -149,42 +170,60 @@
...
-					} else {
-						cb();
-					}
-				}],
-				done);
-
+				(done) => {
+					async.parallel([
+						(cb) => {
+							const webpackAdaptersConfig = require("./build/webpack/webpack.adapters");
```
Comment (NinjaManatee) [inline gulpfile.js]: These requires should be at the top of the file to

File: gulpfile.js [hunk 7]
```diff
@@ -193,36 +232,65 @@
...
-						name: "Finsemble"
+						name: "finsemble-react-controls"
 					}, cb)
 				},
-			], done)
+			], done);
 		},
+		copyStatic: () => {
+			const srcDir = path.join(__dirname)
+			let directories = staticDirectories;
```
Comment (NinjaManatee) [inline gulpfile.js]: I think all of these lets could be const

File: gulpfile.js [hunk 8]
```diff
@@ -326,15 +401,43 @@
...
-				watch(path.join(srcPath, "components", "assets", "**", "*"), {}, this.buildSass),
-				watch(path.join(srcPath, "**", "*.css"), { ignoreInitial: true })
-					.pipe(gulp.dest(distPath)),
-				watch(path.join(srcPath, "**", "*.html"), { ignoreInitial: true })
-					.pipe(gulp.dest(distPath))
-			);
+		watchStatic: (done) => {
+			const srcDir = path.join(__dirname)
+			let directories = staticDirectories;
+			directories.forEach((config) => {
```
Comment (NinjaManatee) [inline gulpfile.js]: Can't you just do the forEach on staticDirectories?

File: gulpfile.js [hunk 9]
```diff
@@ -326,15 +401,43 @@
...
+				let src = config.src.map(src => {
```
Comment (NinjaManatee) [inline gulpfile.js]: I think all of the lets in this function could be const

File: server/server.js [hunk 10]
```diff
@@ -13,76 +13,163 @@
 	// #region Imports
 	// NPM
 	const chalk = require("chalk");
+	const errorColor = chalk.red;
```
Comment (NinjaManatee) [inline server/server.js]: Defining variables in the imports section. This should be done after all fo the imports

File: server/server.js [hunk 11]
```diff
@@ -13,76 +13,163 @@
...
+		}
+	}
+
+	const defaultExtensions = {
+
+		/**
+		 * Method called before starting the server.
+		 *
+		 * @param {function} done Function can take one argument; an error message if one occurred.
+					 * @example
```
Comment (NinjaManatee) [inline server/server.js]: Indentation seems off

File: server/server.js [hunk 12]
```diff
@@ -13,76 +13,163 @@
...
+
+					global.host = server.address().address;
+					global.port = server.address().port;
+				});
+
+			app.use(compression());
+			// Sample server root set to "/" -- must align with paths throughout
+			app.use("/", express.static(rootDir, options));
+			// Open up the Finsemble Components,services, and clients
+			app.use("/Finsemble", express.static(moduleDirectory, options));
```
Comment (NinjaManatee) [inline server/server.js]: Should this be "/finsemble"? It was changed to lowercase elsewhere.

File: server/server.js [hunk 13]
```diff
@@ -13,76 +13,163 @@
...
+			// For Assimilation
+			app.use("/hosted", express.static(path.join(__dirname, "..", "hosted"), options));
```
Comment (NinjaManatee) [inline server/server.js]: Does the hosted folder exist in the seed project? If not, it should be removed here and added in an extension when used.

File: server/server.js [hunk 14]
```diff
@@ -13,76 +13,163 @@
...
+
+			// configs/openfin/manifest-local.json and configs/other/server-environment-startup.json
+			// Make the config public
+			app.use("/configs", express.static("./configs", options));
```
Comment (NinjaManatee) [inline server/server.js]: This seems inconsistent. The configs folder is copied to dist, but is hosted from the project root. We should either not copy it, or host it from dist. I'm for the hosting it from dist.

File: server/server.js [hunk 15]
```diff
@@ -13,76 +13,163 @@
...
-			 *		done(e);
-			 *  }
-			 * }
+			 * Fill in the ProxyMap with entries in order to proxy remote components.
+			 * For instance /connect4/blah/mycomponent.html would be proxied to http://connect4.chartiq.com/blah/mycomponent.html
+			 * 
+			 * This proxy further uses the referer http header to redirect assets from within the component.
 			 */
-			post: done => { done(); },
+			let ProxyMap = [
```
Comment (NinjaManatee) [inline server/server.js]: Hast this been added to the documentation for the seed project?

Comment (NinjaManatee) [inline server/server.js]: Should be const?

File: server/server.js [hunk 16]
```diff
@@ -13,76 +13,163 @@
...
-				app.use("/Finsemble", express.static(moduleDirectory, options));
-				// For Assimilation
-				app.use("/hosted", express.static(path.join(__dirname, "..", "hosted"), options));
-
-				// configs/openfin/manifest-local.json and configs/other/server-environment-startup.json
-				// Make the config public
-				app.use("/configs", express.static("./configs", options));
-
-				cb();
+			if (ProxyMap.length) {
```
Comment (NinjaManatee) [inline server/server.js]: I feel like the application of the proxy map belongs outside of the extension. The developer should be able to add to the proxyMap, but it should be processed in code that we control.

File: server/server.js [hunk 17]
```diff
@@ -13,76 +13,163 @@
...
+
+				const { URL } = require("url");
```
Comment (NinjaManatee) [inline server/server.js]: Require should be at the top of the file.

File: server/server.js [hunk 18]
```diff
@@ -113,6 +202,7 @@
 		const STATS_PATH = path.join(__dirname, "./stats.json");
 		//Listens for the first time that the config and the serviceManager are retrieved, and logs output to the console.
 		let notified_config = false, notified_sm = false;
+		let serviceManagerRetrievedTimeout;
 		app.get("/configs/openfin/manifest-local.json", (req, res, next) => {
```
Comment (NinjaManatee) [inline server/server.js]: Should this be a hard coded path? What if they are using a different file name?

File: src-built-in/components/windowTitleBar/src/stores/windowTitleBarStore.js [hunk 19]
```diff
@@ -35,9 +35,12 @@ var Actions = {
 				{ field: "Close.hide", value: FSBLHeader.hideClose ? true : false },
 			]);
 
-			if (FSBL.Clients.WindowClient.title) {
+			// Set by calling WindowClient.setTitle() || from config "foreign.components.Window Manager.title"
+			var title = FSBL.Clients.WindowClient.title || windowTitleBarConfig.title;
```
Comment (NinjaManatee) [inline src-built-in/components/windowTitleBar/src/stores/windowTitleBarStore.js]: Could be const

File: .gitignore [hunk 20]
```diff
@@ -17,6 +17,8 @@ debug.log
 built
 finsemble/
 .webpack-file-cache
+webpack_cache
 .babel_cache
 build/webpack/vendor-manifest.json
-server/stats.json
\ No newline at end of file
+server/stats.json
+package-lock.json
```
[the remark this answers is not in the document]
Comment (thorsent) [reply .gitignore]: Yeah, I gitignored it. It's already caused problems at JPM.

**Comment by kriswest** [comment]
@thorsent @NinjaManatee just noticed this PR is still open and unassigned. Has this been superceded now we're at (seed) v2.5.3?

**Pull request closed without merging**